File: 2/2_cuatrimestre/PDOO/practicas/practica 2/deepSpaceRuby/SpaceStation.rb
require_relative 'Damage'
require_relative 'Weapon'
require_relative 'ShieldBooster'
require_relative 'Hangar'
require_relative 'SpaceStationToUI'
module Deepspace
    class SpaceStation

        attr_reader :ammoPower, :fuelUnits, :weapons, :hangar, :shieldPower, :name, :nMedals, :pendingDamage, :shieldBoosters

        def initialize(n, supplies)
            @MAXFUEL = 100
            @SHIELDLOSSPERUNITSHOT = 0.1
            @name = n
            @ammoPower = supplies.ammoPower
            @fuelUnits = supplies.fuelUnits
            @shieldPower = supplies.shieldPower
            @weapons = Array.new
            @hangar = nil
            @nMedals = 0
            @pendingDamage = nil
            @shieldBoosters = Array.new
            
        end


        def cleanUpMountedItems
            for i in 0..(@weapons.length-1)
                if (@weapons[i].uses == 0)
                    @weapons.delete_at(i)
                end
            end

            for i in 0..(@shieldBoosters.length-1)
                if (@shieldBoosters[i].uses == 0)
                    @shieldBoosters.delete_at(i)
                end
            end
        end

        def discardHangar
            @hangar = nil
        end

        def discardShieldBooster(i)

        end

        def discardShieldBoosterInHangar(i)
            if (@hangar != nil)
                @hangar.removeShieldBooster(i)
            end
        end

        def discardWeapon(i)

        end


        def discardWeaponInHangar(i)
            if (@hangar != nil)
                @hangar.removeWeapon(i)
            end
        end


        def fire
        
        end

        def mountShieldBooster(i)
            if (@hangar != nil)
                aux = @hangar.removeShieldBooster(i)
                if (aux != nil)
                    shieldBoosters << aux
                end
            end
        end


        def mountWeapon(i)
            if (@hangar != nil)
                aux = @hangar.removeWeapon(i)
                if (aux != nil)
                    weapons << aux
                end
            end
        end


        def move
            @fuelUnits -= getSpeed
            if (@fuelUnits < 0)
                @fuelUnits = 0
            end
            
        end

        def protection

        end

        def receiveHangar(h)
            if (@hangar == nil)
                @hangar = h
            end

        end

        def receiveShieldBooster(s)
            if (@hangar != nil)
                @hangar.addShieldBooster(s)
            else
                false
            end


        end

        def receiveShot(shot)

        end


        def receiveSupplies(s)
            @ammoPower += s.ammoPower
            @shieldPower += s.shieldPower
            @fuelUnits += s.fuelUnits
        end


        def receiveWeapon(w)
            if (@hangar != nil)
                @hangar.addWeapon(w)
            else
                false
            end


        end

        def setLoot(loot)

        end

        def getSpeed
            @fuelUnits/@MAXFUEL
        end

        def setPendingDamage(d)
            @pendingDamage = d.adjust(weapons, shieldBoosters)

        end


        def validState
            (@pendingDamage == nil) || (@pendingDamage.hasNoEffect)
        end


        def getUIversion
            return SpaceStationToUI.new(self)
        end

        def to_s 
            return getUIversion().to_s
        end

        
        private

        def cleanPendingDamage
            if (@pendingDamage.hasNoEffect)
                @pendingDamage = nil
            end


        end


        def assignFuelValue(f)
            if (f <= @MAXFUEL)
                @fuelUnits = f
            end

        end


    end
end

File: 2/2_cuatrimestre/PDOO/practicas/practica 3/deepSpace_3/src/deepspace/SpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;


/**
 *
 * @author manbolq
 */
public class SpaceStation {
    private final int MAXFUEL = 100;
    private final float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    private Hangar hangar;
    private ArrayList<ShieldBooster> shieldBoosters;
    private ArrayList<Weapon> weapons;
    private Damage pendingDamage;
    
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    SpaceStation(String n, SuppliesPackage supplies){
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        weapons = new ArrayList<Weapon>();
        nMedals = 0;
        shieldBoosters = new ArrayList<ShieldBooster>();
    }
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    private void assignFuelValue(float f){
        if (f <= MAXFUEL)
            fuelUnits = f;
    }
    
    private void cleanPendingDamage(){
        if (pendingDamage.hasNoEffect())
            pendingDamage = null;
    }
    
    public void cleanUpMountedItems(){
        for (int i = 0; i < weapons.size(); i++){
            if (weapons.get(i).getUses() == 0)
                weapons.remove(i);
        }
        
        for (int i = 0; i < shieldBoosters.size(); i++){
            if (shieldBoosters.get(i).getUses() == 0)
                shieldBoosters.remove(i);
        }
    }
    
    public void discardHangar(){
        hangar = null;
    }
    
    public void discardShieldBooster(int i){
        int size = shieldBoosters.size();
        
        if (i >= 0 && i < size){
            ShieldBooster s = shieldBoosters.remove(i);
            
            if (pendingDamage != null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (hangar != null)
            hangar.removeShieldBooster(i);
    }
    
    public void discardWeapon(int i){
        int size = weapons.size();
        
        if (i >= 0 && i < size){
            Weapon w = weapons.remove(i);
            
            if (pendingDamage != null){
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (hangar != null)
            hangar.removeWeapon(i);
    }
    
    public float fire(){
        int size = weapons.size();
        float factor = 1.0f;
        
        for (Weapon w : weapons)
            factor *= w.useIt();
        
        return ammoPower*factor;
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public String getName(){
        return name;
     
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public float getSpeed(){
        return fuelUnits/MAXFUEL;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public void mountShieldBooster(int i){
        if (hangar != null){
            ShieldBooster aux = hangar.removeShieldBooster(i);
            if (aux != null)
                shieldBoosters.add(aux);
        }
    }
    
    public void mountWeapon(int i){
        if (hangar != null){
            Weapon aux = hangar.removeWeapon(i);
            if (aux != null)
                weapons.add(aux);
        }
    }
    
    public void move(){
        fuelUnits -= getSpeed();
        if (fuelUnits < 0)
            fuelUnits = 0;
    }
    
    
    public float protection(){
        int size = shieldBoosters.size();
        float factor = 1.0f;
        
        for (ShieldBooster s : shieldBoosters)
            factor *= s.useIt();
        
        return shieldPower*factor;
    }
    
    public void receiveHangar(Hangar h){
        if (hangar == null || hangar.getMaxElements() == 0)
            hangar = h;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if (hangar != null)
            return hangar.addShieldBooster(s);
        else
            return false;
    }
    
    public ShotResult receiveShot(float shot){
        float myProtection = protection();
        
        if (myProtection >= shot){
            shieldPower -= SHIELDLOSSPERUNITSHOT*shot;
            if (shieldPower < 0) shieldPower = 0;
            
            return ShotResult.RESIST;
        }
        else{
            shieldPower = 0;
            return ShotResult.DONOTRESIST;
        }
    }
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        fuelUnits += s.getFuelUnits();
    }
    
    public boolean receiveWeapon(Weapon w){
        if (hangar != null)
            return hangar.addWeapon(w);
        else
            return false;
    }
    
    public void setLoot(Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        
        if (h > 0){
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        for (int i = 1; i <= elements; i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for (int i = 1; i <= elements; i++){
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for (int i = 1; i <= elements; i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
        int medals = loot.getNMedals();
        nMedals += medals;
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState(){
        return ((pendingDamage == null) || (pendingDamage.hasNoEffect()));
    }
    
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
    
    
}
